src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2436,7 +2447,168 @@ public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, bool
                 return ref useStatic ? ref IsNullableEnabledForStaticConstructorsAndFields : ref IsNullableEnabledForInstanceConstructorsAndFields;
             }
 
-            private static ImmutableArray<ImmutableArray<FieldOrPropertyInitializer>> ToReadonlyAndFree(ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>> initializers, out int syntaxLength)
+            private static ImmutableArray<ImmutableArray<FieldOrPropertyInitializer.Builder>> ToReadonlyAndFree(ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>> initializers)
+            {
+                if (initializers.Count == 0)
+                {
+                    initializers.Free();
+                    return ImmutableArray<ImmutableArray<FieldOrPropertyInitializer.Builder>>.Empty;
+                }
+
+                var builder = ArrayBuilder<ImmutableArray<FieldOrPropertyInitializer.Builder>>.GetInstance(initializers.Count);
+                foreach (ArrayBuilder<FieldOrPropertyInitializer.Builder> group in initializers)
+                {
+                    builder.Add(group.ToImmutableAndFree());
+                }
+
+                initializers.Free();
+                return builder.ToImmutableAndFree();
+            }
+
+            public void Free()
+            {
+                NonTypeNonIndexerMembers.Free();
+
+                foreach (var group in StaticInitializers)
+                {
+                    group.Free();
+                }
+                StaticInitializers.Free();
+
+                foreach (var group in InstanceInitializers)
+                {
+                    group.Free();
+                }
+                InstanceInitializers.Free();
+
+                IndexerDeclarations.Free();
+            }
+
+            internal void AddOrWrapTupleMembers(SourceMemberContainerTypeSymbol type)
+            {
+                this.NonTypeNonIndexerMembers = type.AddOrWrapTupleMembers(this.NonTypeNonIndexerMembers.ToImmutableAndFree());
+            }
+        }
+
+        protected sealed class DeclaredMembersAndInitializers
+        {
+            public readonly ImmutableArray<Symbol> NonTypeNonIndexerMembers;
+            public readonly ImmutableArray<ImmutableArray<FieldOrPropertyInitializer>> StaticInitializers;
+            public readonly ImmutableArray<ImmutableArray<FieldOrPropertyInitializer.Builder>> InstanceInitializers;
+            public readonly ImmutableArray<SyntaxReference> IndexerDeclarations;
+            public readonly RecordDeclarationSyntax? RecordDeclarationWithParameters;
+            public readonly int StaticInitializersSyntaxLength;
```

Comment:
I think we might want to calculate this at the end, when we do similar calculations for instance initializers. Also, will be able to reduce the size of this instance somewhat.

Reply:
We can (see commit 2) but then we need two sets of ToReadonly methods. One that operates on ArrayBuilder and one that ImmutableArray.
I purposefully went the other direction. The trade-off is storing one less integer in this data structure versus making code more complex.

---
In reply to: [562781553](https://github.com/dotnet/roslyn/pull/50650#discussion_r562781553) [](ancestors = 562781553)

Reply:
> We can (see commit 2) but then we need two sets of ToReadonly methods. One that operates on ArrayBuilder and one that ImmutableArray.
I purposefully went the other direction. The trade-off is storing one less integer in this data structure versus making code more complex.

I think the fact that we try to pre-calculate some syntax lengths that can easily be calculated later from information we already creates an unnecessary complexity. I suggest getting rig of that length information here and in FieldOrPropertyInitializer.  It looks like that length information is used only in one function and we can easily calculate it on demand.  

---
In reply to: [562797338](https://github.com/dotnet/roslyn/pull/50650#discussion_r562797338) [](ancestors = 562797338,562781553)